File: packages/go/slop-ai/validate_params.go
package slop

import (
	"encoding/json"
	"fmt"
)

// ValidateParams checks `params` against the affordance's JSON Schema and
// returns a human-readable error string on failure (empty string on success).
// Mirrors the TS/Python implementations so the `invalid_params` error code is
// reliable across SDKs.
//
// Supported schema subset:
//   - type: object|string|number|integer|boolean|array|null
//   - required (for objects)
//   - properties (for objects)
//   - items (for arrays)
//   - enum
func ValidateParams(schema map[string]any, params any) string {
	if len(schema) == 0 {
		return ""
	}
	return validateParamsImpl(schema, params, "params")
}

func validateParamsImpl(schema map[string]any, value any, path string) string {
	if raw, ok := schema["enum"]; ok {
		if list, ok := raw.([]any); ok {
			matched := false
			for _, opt := range list {
				if deepEqual(opt, value) {
					matched = true
					break
				}
			}
			if !matched {
				b, _ := json.Marshal(list)
				return fmt.Sprintf("%s must be one of %s", path, string(b))
			}
		}
	}

	t, _ := schema["type"].(string)
	switch t {
	case "object":
		obj, ok := value.(map[string]any)
		if !ok {
			return fmt.Sprintf("%s must be an object", path)
		}
		for _, key := range requiredKeys(schema["required"]) {
			if _, present := obj[key]; !present {
				return fmt.Sprintf("%s.%s is required", path, key)
			}
		}
		if props, ok := schema["properties"].(map[string]any); ok {
			for key, v := range obj {
				if ps, ok := props[key].(map[string]any); ok {
					if err := validateParamsImpl(ps, v, path+"."+key); err != "" {
						return err
					}
				}
			}
		}
		return ""
	case "array":
		arr, ok := value.([]any)
		if !ok {
			return fmt.Sprintf("%s must be an array", path)
		}
		if items, ok := schema["items"].(map[string]any); ok {
			for i, item := range arr {
				if err := validateParamsImpl(items, item, fmt.Sprintf("%s[%d]", path, i)); err != "" {
					return err
				}
			}
		}
		return ""
	case "string":
		if _, ok := value.(string); ok {
			return ""
		}
		return fmt.Sprintf("%s must be a string", path)
	case "number":
		switch value.(type) {
		case float64, float32, int, int64, int32:
			return ""
		}
		return fmt.Sprintf("%s must be a number", path)
	case "integer":
		switch v := value.(type) {
		case int, int32, int64:
			return ""
		case float64:
			if v == float64(int64(v)) {
				return ""
			}
		}
		return fmt.Sprintf("%s must be an integer", path)
	case "boolean":
		if _, ok := value.(bool); ok {
			return ""
		}
		return fmt.Sprintf("%s must be a boolean", path)
	case "null":
		if value == nil {
			return ""
		}
		return fmt.Sprintf("%s must be null", path)
	}
	return ""
}

// requiredKeys normalizes schema["required"], which may be []string when the
// schema was built from Go via normalizeParams or []any when decoded from JSON.
func requiredKeys(raw any) []string {
	switch v := raw.(type) {
	case []string:
		out := make([]string, 0, len(v))
		for _, k := range v {
			if k != "" {
				out = append(out, k)
			}
		}
		return out
	case []any:
		out := make([]string, 0, len(v))
		for _, r := range v {
			if k, ok := r.(string); ok && k != "" {
				out = append(out, k)
			}
		}
		return out
	}
	return nil
}

func deepEqual(a, b any) bool {
	if a == nil && b == nil {
		return true
	}
	aj, errA := json.Marshal(a)
	bj, errB := json.Marshal(b)
	if errA != nil || errB != nil {
		return false
	}
	return string(aj) == string(bj)
}

File: packages/rust/slop-ai/src/validate_params.rs
//! Minimal JSON Schema validator for affordance invoke params.
//!
//! Mirrors the TS/Python/Go implementations so the `invalid_params` error
//! code is reliable across SDKs. Supported subset:
//!
//! * `type`: object | string | number | integer | boolean | array | null
//! * `required` (for objects)
//! * `properties` (for objects)
//! * `items` (for arrays)
//! * `enum`

use serde_json::Value;

/// Returns `None` on success or a human-readable error message.
pub fn validate_params(schema: Option<&Value>, params: &Value) -> Option<String> {
    let Some(schema) = schema else { return None };
    if !schema.is_object() {
        return None;
    }
    validate(schema, params, "params")
}

fn validate(schema: &Value, value: &Value, path: &str) -> Option<String> {
    if let Some(options) = schema.get("enum").and_then(|v| v.as_array()) {
        if !options.iter().any(|opt| opt == value) {
            return Some(format!(
                "{path} must be one of {}",
                serde_json::to_string(options).unwrap_or_default()
            ));
        }
    }

    match schema.get("type").and_then(|v| v.as_str()) {
        Some("object") => {
            let Some(obj) = value.as_object() else {
                return Some(format!("{path} must be an object"));
            };
            if let Some(required) = schema.get("required").and_then(|v| v.as_array()) {
                for key in required {
                    if let Some(k) = key.as_str() {
                        if !obj.contains_key(k) {
                            return Some(format!("{path}.{k} is required"));
                        }
                    }
                }
            }
            if let Some(props) = schema.get("properties").and_then(|v| v.as_object()) {
                for (key, v) in obj {
                    if let Some(prop_schema) = props.get(key) {
                        let sub_path = format!("{path}.{key}");
                        if let Some(err) = validate(prop_schema, v, &sub_path) {
                            return Some(err);
                        }
                    }
                }
            }
            None
        }
        Some("array") => {
            let Some(arr) = value.as_array() else {
                return Some(format!("{path} must be an array"));
            };
            if let Some(items) = schema.get("items") {
                for (i, item) in arr.iter().enumerate() {
                    let sub_path = format!("{path}[{i}]");
                    if let Some(err) = validate(items, item, &sub_path) {
                        return Some(err);
                    }
                }
            }
            None
        }
        Some("string") => {
            if value.is_string() {
                None
            } else {
                Some(format!("{path} must be a string"))
            }
        }
        Some("number") => {
            if value.is_number() {
                None
            } else {
                Some(format!("{path} must be a number"))
            }
        }
        Some("integer") => {
            if value.is_i64() || value.is_u64() {
                None
            } else if let Some(f) = value.as_f64() {
                if f.fract() == 0.0 {
                    None
                } else {
                    Some(format!("{path} must be an integer"))
                }
            } else {
                Some(format!("{path} must be an integer"))
            }
        }
        Some("boolean") => {
            if value.is_boolean() {
                None
            } else {
                Some(format!("{path} must be a boolean"))
            }
        }
        Some("null") => {
            if value.is_null() {
                None
            } else {
                Some(format!("{path} must be null"))
            }
        }
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn empty_schema_ok() {
        assert_eq!(validate_params(None, &json!({"x": 1})), None);
        assert_eq!(
            validate_params(Some(&json!({"type": "object"})), &json!({})),
            None
        );
    }

    #[test]
    fn required_missing() {
        let schema = json!({"type": "object", "required": ["body"]});
        let err = validate_params(Some(&schema), &json!({}));
        assert!(err.unwrap().contains("body"));
    }

    #[test]
    fn wrong_type() {
        let schema = json!({"type": "object", "properties": {"count": {"type": "integer"}}});
        let err = validate_params(Some(&schema), &json!({"count": "x"}));
        assert!(err.unwrap().contains("count"));
    }

    #[test]
    fn array_items() {
        let schema = json!({
            "type": "object",
            "properties": {"tags": {"type": "array", "items": {"type": "string"}}}
        });
        assert_eq!(
            validate_params(Some(&schema), &json!({"tags": ["a", "b"]})),
            None
        );
        let err = validate_params(Some(&schema), &json!({"tags": ["a", 2]})).unwrap();
        assert!(err.contains("tags[1]"));
    }
}
